components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java

```diff
@@ -0,0 +1,459 @@
+					ByteArrayInputStream in = new ByteArrayInputStream(data);
+					ObjectInputStream is = new ObjectInputStream(in);
+					DefaultExchangeHolder removedHolder = (DefaultExchangeHolder) is.readObject();
+					ByteArrayOutputStream bos = new ByteArrayOutputStream();
+					ObjectOutputStream oos = new ObjectOutputStream(bos);
+					oos.writeObject(removedHolder);
+					oos.flush();
+					data = bos.toByteArray();
+					transaction
+							.If(new Cmp(ByteSequence.from(String.format("%s/%s", prefixName, key).getBytes()),
+									Cmp.Op.EQUAL,
+									CmpTarget.value(
+											ByteSequence.from(String.format("%s/%s", prefixName, key).getBytes()))))
+							.Then(Op.delete(ByteSequence.from(String.format("%s/%s", prefixName, key).getBytes()),
+									DeleteOption.DEFAULT),
+									Op.put(ByteSequence
+											.from(String.format("%s/%s", persistencePrefixName, key).getBytes()),
+											ByteSequence.from(data), PutOption.DEFAULT))
+							.commit();
+					LOG.trace("Removed an exchange with ID {} for key {} in a thread-safe manner.",
+							exchange.getExchangeId(), key);
+					LOG.trace(
+							"Put an exchange with ID {} for key {} into a recoverable storage in a thread-safe manner.",
+							exchange.getExchangeId(), key);
+				} catch (Throwable throwable) {
+					throw new RuntimeException(throwable);
+				}
+			} else {
+				CompletableFuture<DeleteResponse> completableDeleteResponse = kvClient
+						.delete(ByteSequence.from(String.format("%s/%s", prefixName, key).getBytes()));
+				try {
+					completableDeleteResponse.get();
+				} catch (InterruptedException | ExecutionException e) {
+					LOG.error(e.getMessage(), e);
+				}
+			}
+		}
+	}
+
+	@Override
+	public void confirm(CamelContext camelContext, String exchangeId) {
+		LOG.trace("Confirming an exchange with ID {}.", exchangeId);
+		if (useRecovery) {
+			CompletableFuture<DeleteResponse> completableDeleteResponse = kvClient
+					.delete(ByteSequence.from(String.format("%s/%s", persistencePrefixName, exchangeId).getBytes()));
+			try {
+				completableDeleteResponse.get();
+			} catch (InterruptedException | ExecutionException e) {
+				LOG.error(e.getMessage(), e);
+			}
+		}
+	}
+
+	@Override
+	public Set<String> getKeys() {
+		CompletableFuture<GetResponse> completableGetResponse = kvClient.get(ByteSequence.from(prefixName.getBytes()),
+				GetOption.newBuilder().withRange(ByteSequence.from(prefixName.getBytes())).build());
+		Set<String> scanned = Collections.unmodifiableSet(new TreeSet<>());
+		try {
+			GetResponse getResponse = completableGetResponse.get();
+			Set<String> keys = new TreeSet<>();
+			getResponse.getKvs().forEach(kv -> keys.add(new String(kv.getKey().getBytes())));
+			scanned = Collections.unmodifiableSet(keys);
+		} catch (InterruptedException | ExecutionException e) {
+			LOG.error(e.getMessage(), e);
+		}
+		return scanned;
+	}
+
+	@Override
+	protected void doStart() throws Exception {
+		if (maximumRedeliveries < 0) {
```

Comment:
Validation can be moved to `onInit` to check correctness of the parameters early